### DISCUSSION TITLE
Triggering async function on button click

### DISCUSSION THREAD

**Discussion by vascoferreira25** [emilk/egui#634] [Q&A] [answered]
I want to trigger a download when I click on a button, but the download function is async using tokio.

I have something like this inside the `event_loop.run`

```
if ui.button("Download").clicked() {
    for l in links.clone() {
        video_downloader.add_video(l.clone());
    }

    video_downloader.download().await;
}
```

But I'm getting the error that it must be inside an async block.
How should I fix this?

**Comment by coderedart** [comment, marked as answer]
If you want the application to stop responding until download completes, you can use spawn blocking method of tokio which will download the links but hang until it completes. But a better way would be to send the downloads to tokio which should be running in another thread. Tokio recently updated the guide for this purpose. https://tokio.rs/tokio/topics/bridging

You spawn tokio on another thread, and use the handle or a channel to send tasks like download video to tokio. The main thread and event loop will continue normally because downloads are happening in another thread.

**Reply by vascoferreira25** [reply]
Thanks. It worked by following the tokio/bridging section. Can't mark as answer for an unknown reason.

**Reply by Walther** [reply]
Would it be possible to have an official `egui` example for this? I'm guessing spawning an async task that does not block the rendering of the GUI (and can e.g. show a progress bar instead) would be a pretty common need, for pretty much anything with longer-running or resource-intensive tasks. Thanks!

Let me know if you want an issue created for this

**Reply by emilk** [reply]
I think an example would be great - can you make a PR?

**Reply by NiceneNerd** [reply]
I would be very curious to see such a thing. I'm little experienced with threading in Rust and have been banging my head against the wall all day on this issue.

**Reply by vascoferreira25** [reply]
I'm really ashamed to say this but although I can trigger an async function by spawning a thread, the window stays blocked until the function that thread is over. I simply choose to ignore creating a perfect gui and just fucused on making it work. I'm sorry.

The program I wrote is the following: [repo](https://github.com/vascoferreira25/arquivos_rtp_downloader)

When I click on the download button, it starts a new thread and all the downloads start. However, the main thread blocks until the new thread finishes. You can easily check why it blocks in [this file](https://github.com/vascoferreira25/arquivos_rtp_downloader/blob/aa2bb411b47722a3c4bbe04ff8eec1d9e205323b/src/gui/main.rs#L80):

```rust
ui.with_layout(egui::Layout::top_down(egui::Align::Center), |ui| {
    if ui.button("Adicionar link").clicked() {
        links.push(link.clone());
        link = String::new();
    }
    
    if ui.button("Limpar Lista").clicked() {
        links = Vec::new();
    }

    if ui.button("Download").clicked() {
        for l in links.clone() {
            video_downloader.add_video(l.clone());
        }

        runtime.block_on(video_downloader.download());
    }
});
```

However, I can't use the function `runtime.spawn()` as it triggers an error: 

> argument requires that `'1` must outlive `'static` 

I don't know what to do now.

**Reply by NiceneNerd** [reply]
I figured out a way to get it to work using messages. The main update code (`send` and `recv` being members of the app struct):
```rust
            if ui.button("Process").clicked() {
                *show_busy = true;
                let folderc = folder.clone();
                let send = send.clone();
                std::thread::spawn(move || {
                    send.send(process_maps(&folderc)).unwrap();
                });
            }
            if let Ok(res) = recv.try_recv() {
                *show_busy = false;
                match res {
                    Ok(objs) => *objects = objs,
                    Err(e) => {
                        *show_error = true;
                        *error = Some(e.to_string());
                    }
                }
            }
```
You can see more here: https://github.com/NiceneNerd/hypostasis/blob/master/src/app.rs

**Reply by ChrisRega** [reply]
I did it like this for my example-blog-reader: https://github.com/ChrisRega/example-blog-client
On polling for readyness it starts the async updater function for either a generic single value or a vec of single values which can be read out each frame as slice. It has some basic error handling and consists of pretty generic primitives.

**Comment by emilk** [comment]
I made a simple example of triggering a concurrent download using the [`poll-promise`](https://docs.rs/poll-promise) crate: https://github.com/emilk/egui/blob/master/eframe/examples/download_image.rs

If you are using `tokio`, you can use [`poll_promise::Promise::spawn_async`](https://docs.rs/poll-promise/latest/poll_promise/struct.Promise.html#method.spawn_async).

**Reply by jellybobbin** [reply]
It doesn't work properly

**Reply by emilk** [reply]
It works for me. If you want help, you need to provide more info than "doesn't work properly".

**Reply by jellybobbin** [reply]
Git pull and then run it is normal, but replace the egui with the crates.io's version, it can't work. I think it's the reason for the problem of the egui version. Thank you...

**Reply by lfrancke** [reply]
I know that this discussion is relatively old and I'm sorry but I thought it might fit here instead of a new thread.

I've tried your example - thank you for that! And it works fine. I then ventured on the seemingly simple quest to extend that to use the `spawn_async` variant you mentioned and failed.

My attempt was to create a `Runtime` and keep it as part of my App but anytime I do that I get the error message that the Promise sender has already been dropped. I admit that I don't fully understand the async bits yet.

I did see this [example](https://github.com/parasyte/egui-tokio-example/blob/main/src/main.rs) from another [discussion](https://github.com/emilk/egui/discussions/521) and while that works I tried to understand better and I tried to not have a separate thread doing nothing but holding the runtime.

```rust
pub struct TemplateApp {
    promise: Option<Promise<String>>>,
    rt: runtime::Runtime,
}

impl Default for TemplateApp {
    fn default() -> Self {
        Self {
            promise: None,
            rt: runtime::Builder::new_multi_thread()
                .enable_all()
                .build()
                .unwrap(),
        }
    }
}

impl eframe::App for TemplateApp {
    fn update(&mut self, ctx: &egui::Context, _frame: &mut eframe::Frame) {
        let promise = self.albums.get_or_insert_with(|| {
            let _guard = self.rt.enter();
            let promise = Promise::spawn_async(async move {
             foobar.await
            });
            promise
        });
...

```

Error: `The Promise Sender was dropped`
I believe this has probably something to do with the guard dropping too early (the tokio docs just mention "runtime context" without going into detail) and I believe that's probably why the other example works because said context is kept alive.

i just wondered if you had an example on how `spawn_async` would be/can be used with `tokio` in `egui`, thank you.

**Reply by ChrisRega** [reply]
I just used tokio's `macro` feature to wrap everything in main into async already, then you don't need to manage the runtime yourself. Then you can just use `tokio::spawn` wherever you want. You can see an example in this line: 
[code](https://github.com/ChrisRega/example-blog-client/blob/main/src/main.rs#L86)
The crate underneath is `lazy_async_promise`, especially the `ImmediateValuePromise` from it, which spawns your future directly for you, see: [docs](https://docs.rs/lazy_async_promise/0.3.1/lazy_async_promise/struct.ImmediateValuePromise.html)
Maybe this helps :)

**Reply by lfrancke** [reply]
Thank you @ChrisRega - I did see your example and I tried it before!

I didn't phrase it very well: I do have a working solution. At this point it's about trying to get a better understanding why the _other_ solution doesn't work and what'd be required to get it working.

**Reply by JiveyGuy** [reply]
> I made a simple example of triggering a concurrent download using the [`poll-promise`](https://docs.rs/poll-promise) crate: https://github.com/emilk/egui/blob/master/eframe/examples/download_image.rs
> 
> If you are using `tokio`, you can use [`poll_promise::Promise::spawn_async`](https://docs.rs/poll-promise/latest/poll_promise/struct.Promise.html#method.spawn_async).

This link is dead now :(

**Reply by coderedart** [reply]
https://github.com/emilk/egui/tree/master/examples/download_image

**Comment by wiseaidev** [comment]
It seems like threads/tasks are not supported on wasm/web yet, available only for native. I tried both spawning threads/tasks, didn't work:

![thread spawn](https://github.com/emilk/egui/assets/62179149/f8a06a22-b958-4aa4-a71f-221312757b74)
![tokio](https://github.com/emilk/egui/assets/62179149/38ee8f7b-26f6-4a9b-b3ac-31be10d7082f)

If anyone knows how to do so on wasm, your reply would be much appreciated!

Edit: [`poll_promise::Promise::spawn_thread`](https://docs.rs/poll-promise/latest/poll_promise/struct.Promise.html#method.spawn_thread) is non wasm as well

**Reply by c-git** [reply]
Yes if you need to support WASM async is a bit harder but not unachievable. Here is an [example app](https://github.com/woelper/egui_pick_file) from the [readme](https://github.com/emilk/egui?tab=readme-ov-file#how-to-i-create-a-file-dialog) that shows how to do async for both native and WASM. (The async task is file load / save)

I can't remember if that one uses poll_promise but [this one](https://github.com/c-git/log-viewer/) does use it for WASM (well it uses poll_promise for both).
